### PR TITLE
[blog] flink on zeppelin - part2

### PR DESCRIPTION
**Pull request by zjffdu**
This is part2 of flink on zeppelin blog.

### CODE REVIEW

**Comment by zjffdu** [comment]
@morsapaes @MarkSfik This is part2 of flink on zeppelin

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+title:  "Flink on Zeppelin Notebooks for Interactive Data Analysis - Part 2"
+date:   2020-05-25T08:00:00.000Z
+categories: ecosystem
+authors:
+- zjffdu:
+  name: "Jeff Zhang"
+  twitter: "zjffdu"
+---
+
+In the last post, I introduce the basics of Flink on Zeppelin and how to do Streaming ETL. This is part-2 where I would talk about how to
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
In a previous post, we introduced the basics of Flink on Zeppelin and how to do Streaming ETL. In this second part of the "Flink on Zeppelin" series of posts, I will share how to 
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+date:   2020-05-25T08:00:00.000Z
+categories: ecosystem
+authors:
+- zjffdu:
+  name: "Jeff Zhang"
+  twitter: "zjffdu"
+---
+
+In the last post, I introduce the basics of Flink on Zeppelin and how to do Streaming ETL. This is part-2 where I would talk about how to 
+do streaming data visualization via Flink on Zeppelin and how to use flink UDF in Zeppelin.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
perform streaming data visualization via Flink on Zeppelin and how to use Apache Flink UDFs in Zeppelin. 
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  name: "Jeff Zhang"
+  twitter: "zjffdu"
+---
+
+In the last post, I introduce the basics of Flink on Zeppelin and how to do Streaming ETL. This is part-2 where I would talk about how to 
+do streaming data visualization via Flink on Zeppelin and how to use flink UDF in Zeppelin. 
+
+# Streaming Data Visualization
+
+In Zeppelin, you can build a realtime streaming dashboard without writing any line of javascript/html/css code.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
With [Zeppelin](https://zeppelin.apache.org/), you can build a real time streaming dashboard without writing any line of javascript/html/css code.
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+Overall Zeppelin supports 3 kinds of streaming data analytics:
+* Single
+* Update
+* Append
+
+### Single Mode
+Single mode is for the case when the result of sql statement is always one row, such as the following example.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
Single mode is used for cases when the result of a SQL statement is always one row, such as the following example. 
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+In Zeppelin, you can build a realtime streaming dashboard without writing any line of javascript/html/css code.
+Overall Zeppelin supports 3 kinds of streaming data analytics:
+* Single
+* Update
+* Append
+
+### Single Mode
+Single mode is for the case when the result of sql statement is always one row, such as the following example. 
+The output format is HTML, and you can specify paragraph local property template for the final output content template.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
The output format is translated in HTML, and you can specify a paragraph local property template for the final output content template. 
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+In Zeppelin, you can build a realtime streaming dashboard without writing any line of javascript/html/css code.
+Overall Zeppelin supports 3 kinds of streaming data analytics:
+* Single
+* Update
+* Append
+
+### Single Mode
+Single mode is for the case when the result of sql statement is always one row, such as the following example. 
+The output format is HTML, and you can specify paragraph local property template for the final output content template. 
+And you can use {i} as placeholder for the ith column of result.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]: I am not sure what "ith" maybe you want to ad {i}th if you want to say 9th or 8th? 
It should also be "column of **the** result"

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_single_mode.gif" width="80%" alt="Single Mode"/>
+</center>
+
+### Update Mode
+Update mode is suitable for the case when the output is more than one rows,
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
Update mode is suitable for the cases when the output format is more than one row, 
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+The output format is HTML, and you can specify paragraph local property template for the final output content template. 
+And you can use {i} as placeholder for the ith column of result.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_single_mode.gif" width="80%" alt="Single Mode"/>
+</center>
+
+### Update Mode
+Update mode is suitable for the case when the output is more than one rows, 
+and always will be updated continuously. Here’s one example where we use group by.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
and will always be continuously updated. Here’s one example where we use group by.
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_update_mode.gif" width="80%" alt="Update Mode"/>
+</center>
+
+### Append Mode
+Append mode is suitable for the scenario where output data is always appended.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
Append mode is suitable for the cases when the output data is always appended. 
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+Update mode is suitable for the case when the output is more than one rows, 
+and always will be updated continuously. Here’s one example where we use group by.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_update_mode.gif" width="80%" alt="Update Mode"/>
+</center>
+
+### Append Mode
+Append mode is suitable for the scenario where output data is always appended. 
+E.g. the following example which use tumble window.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
For instance, the example below uses a tumble window.
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_append_mode.gif" width="80%" alt="Append Mode"/>
+</center>
+
+# UDF
+
+SQL is powerful, especially in expressing data flow. But most of time, you need to handle complicated business logic that can not be expressed by SQL.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
SQL is a very powerful language, especially in expressing data flow. But most of the time, you need to handle complicated business logic that cannot be expressed by SQL.
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+In these cases, you will need UDF (user defined function). In Zeppelin, you can write Scala, Python UDF, and also import Scala, Python and Java UDF.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
In these cases UDFs (user-defined functions) come particularly handy. In Zeppelin, you can write Scala or Python UDFs, while you can also import Scala, Python and Java UDFs.
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+Here're 2 examples of Scala and Python UDF.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
Here are 2 examples of Scala and Python UDFs:
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+class PythonUpper(ScalarFunction):
+def eval(self, s):
+ return s.upper()
+
+bt_env.register_function("python_upper", udf(PythonUpper(), DataTypes.STRING(), DataTypes.STRING()))
+
+```
+
+After you define the UDF, you can use them directly in SQL. e.g.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
After you define the UDFs, you can use them directly in SQL. e.g.
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+* Use Python UDF in SQL
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_python_udf.png" width="100%" alt="Python UDF"/>
+</center>
+
+# Summary
+
+In this post, I show you how to do streaming data visualization via Flink on Zeppelin and how to use UDF.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
In this post, we explained how to perform streaming data visualization via Flink on Zeppelin and how to use UDFs. 
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+* Use Python UDF in SQL
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_python_udf.png" width="100%" alt="Python UDF"/>
+</center>
+
+# Summary
+
+In this post, I show you how to do streaming data visualization via Flink on Zeppelin and how to use UDF. 
+Besides that, you can do more in Zeppelin with Flink, such as batch processing, hive integration and etc.
```
Comment (MarkSfik) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
Besides that, you can do more in Zeppelin with Flink, such as batch processing, hive integration and more.
```

**Review by MarkSfik — COMMENTED** [review]
Thanks for sharing the second part of the story @zjffdu 
I left some suggestions and comments here.

**Comment by zjffdu** [comment]
Thanks for the review @MarkSfik , I have updated the post

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+In a previous post, we introduced the basics of Flink on Zeppelin and how to do Streaming ETL. In this second part of the "Flink on Zeppelin" series of posts, I will share how to 
+perform streaming data visualization via Flink on Zeppelin and how to use Apache Flink UDFs in Zeppelin. 
+
+# Streaming Data Visualization
+
+With [Zeppelin](https://zeppelin.apache.org/), you can build a real time streaming dashboard without writing any line of javascript/html/css code.
+Overall Zeppelin supports 3 kinds of streaming data analytics:
+* Single
+* Update
+* Append
```
Comment (morsapaes) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion

Overall, Zeppelin supports 3 kinds of streaming data analytics:

* Single Mode
* Update Mode
* Append Mode
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+The output format is translated in HTML, and you can specify a paragraph local property template for the final output content template. 
+And you can use `{i}` as placeholder for the {i}th column of the result.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_single_mode.gif" width="80%" alt="Single Mode"/>
+</center>
+
+### Update Mode
+Update mode is suitable for the cases when the output format is more than one row, 
+and will always be continuously updated. Here’s one example where we use group by.
```
Comment (morsapaes) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
and will always be continuously updated. Here’s one example where we use ``GROUP BY``.
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+class PythonUpper(ScalarFunction):
+def eval(self, s):
+ return s.upper()
+
+bt_env.register_function("python_upper", udf(PythonUpper(), DataTypes.STRING(), DataTypes.STRING()))
+
+```
+
+After you define the UDFs, you can use them directly in SQL. e.g.
```
Comment (morsapaes) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
After you define the UDFs, you can use them directly in SQL:
```

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+* Use Python UDF in SQL
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_python_udf.png" width="100%" alt="Python UDF"/>
+</center>
+
+# Summary
+
+In this post, we explained how to perform streaming data visualization via Flink on Zeppelin and how to use UDFs. 
+Besides that, you can do more in Zeppelin with Flink, such as batch processing, hive integration and more.
```
Comment (morsapaes) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
Besides that, you can do more in Zeppelin with Flink, such as batch processing, Hive integration and more.
```

**Review by morsapaes — COMMENTED** [review]
Thanks for this follow-up blogpost, @zjffdu! I think we can leave some time between publishing the blogposts instead of publishing both at the same time, what do you think?

**Comment by zjffdu** [comment]
Thanks @morsapaes for the review, I have updated post, publishing separately at different time make sense to me. How about publishing part-1 this week, and part-2 next week  ?

**Comment by morsapaes** [comment]
Sounds good to me, @zjffdu!

File: _posts/2020-05-25-flink-on-zeppelin-part2.md [hunk 21]
```diff
@@ -0,0 +1,109 @@
+---
+layout: post
+title:  "Flink on Zeppelin Notebooks for Interactive Data Analysis - Part 2"
+date:   2020-05-25T08:00:00.000Z
```
Comment (morsapaes) [inline _posts/2020-05-25-flink-on-zeppelin-part2.md]:
```suggestion
date:   2020-06-23T08:00:00.000Z
```

Comment (morsapaes) [reply _posts/2020-05-25-flink-on-zeppelin-part2.md]: @zjffdu, will try to get this one out next week.

**Comment by zjffdu** [comment]
Thanks @morsapaes

**Review by carp84 — APPROVED** [review]
+1, thanks for the contribution @zjffdu and @morsapaes !

Will merge this with some minor adjustments to coordinate the file name of the blog markdown with the date in blog header, as well as image directory name, to make sure the blog content appears in the right place on our website.

**Comment by zjffdu** [comment]
@carp84 Thanks for helping merging this.

**Pull request closed without merging**